### ISSUE TITLE
How to change sshd default port

### ISSUE THREAD

**Issue by salesgroup** [coreos/fedora-coreos-tracker#396]
We are starting the migration from CoreOS to Fedora CoreOS.
No documentation (from Fedora CoreOS) about changing the SSH ports.
So we tried this:
https://coreos.com/os/docs/latest/customizing-sshd.html#changing-the-sshd-port-1

file: /etc/systemd/system/sshd.socket.d/10-sshd-listen-ports.conf
> [Socket]
> ListenStream=
> ListenStream=12221
> 

But we got this error:
systemd[14685]: sshd.socket: Failed to create listening socket ([::]:12221): Permission denied

Seems related with SElinux.

How can we change the sshd port?

**Comment by bgilbert** [comment]
CoreOS Container Linux used systemd socket activation for sshd, so the listening port was configured via systemd.  Fedora CoreOS runs sshd in the traditional way, as a service that's started automatically, so you'd configure the port in `/etc/ssh/sshd_config`.

This FCC will do half of the work:
```
variant: fcos
version: 1.0.0
storage:
  files:
    - path: /etc/ssh/sshd_config
      append:
        - inline: |
            Port 2222
```

But in addition, `sshd_config` has this comment:

```
# If you want to change the port on a SELinux system, you have to tell
# SELinux about this change.
# semanage port -a -t ssh_port_t -p tcp #PORTNUMBER
```

And we don't ship `semanage`.

**Comment by salesgroup** [comment]
I had already tried this but it gave the same type of error.

I think that i've to configure SELinux to give access to sshd to this port.
How can i do that?

**Comment by jlebon** [comment]
I think today this requires modifying the binary policy, which [is going to be an issue](https://github.com/coreos/fedora-coreos-tracker/issues/368). I'm not sure if there's a way to only change the runtime value so we can play the same "modify on boot" trick. We can look into that, but we should really get together with the SELinux folks to get this fixed properly.

**Comment by salesgroup** [comment]
Since "semanage" is not available, how can we do it after server boot?

**Comment by dustymabe** [comment]
@salesgroup I don't think we have a good answer for you right now. We'd have to figure out how to do what semanage does without running semanage. We have some gaps here on how to manipulate selinux policy in a way that makes more sense for Fedora CoreOS. Ideally any changes to selinux policy would be able to be configured via plain text files that are laid down using Ignition.

Possible workarounds (all are bad):
- You might be able to bindmount things in and run `semanage` from a container
- You may choose to just package layer  the `policycoreutils-python-utils` rpm (though see #400)
- A really hacky answer may be to run socat or something to redirect traffic from external port 12221 traffic to port 22 locally and block port external 22 using a firewall config

None of those are great answers. We [have a similar problem](https://github.com/coreos/fedora-coreos-tracker/issues/368) with selinux booleans. As a group we need to talk to the SELinux team and find a path forward for the "lay down config via plain text files" approach.

**Comment by dustymabe** [comment]
@salesgroup do any of those proposed workaround unblock you?

**Comment by icedream** [comment]
I have also stumbled across this issue since I need to move the SSH server away from port 22 - it was already in use for Git SSH on a legacy CoreOS installation.

I decided to follow the workaround of running `semanage` from a container. For this I wrote a script that is installed into CoreOS during Ignition phase and is then run by systemd right before the SSH server starts up via a service dropin. I also disabled the socket activation to let sshd directly listen on the port, but I think that doesn't matter here.

I have posted the actual script, systemd dropin and resulting ignition config parts to [this gist](https://gist.github.com/icedream/75135f63f433ec52d652c7245dd17e30).

**Comment by bgilbert** [comment]
@icedream  Thanks for the example!  Fedora CoreOS, unlike Container Linux, doesn't use socket activation for sshd, so enabling `sshd.service` / masking `sshd.socket` shouldn't be necessary.

**Comment by remoe** [comment]
I wrote for completeness the hacky way (idea from: https://github.com/coreos/fedora-coreos-tracker/issues/396#issuecomment-592028415 ) using proxying 2222 to 22:

```ini
    - name: ssh-proxy.service
      enabled: true
      contents: |
        [Unit]
        Description=SSH Proxy
        Wants=network-online.target
        After=network-online.target

        [Install]
        WantedBy=multi-user.target

        [Service]
        Type=notify
        KillMode=mixed
        TimeoutStartSec=0
        Restart=always
        RestartSec=5s
        ExecStart=/usr/bin/socat TCP4-LISTEN:2222,reuseaddr,fork TCP4:127.0.0.1:22
```

**Comment by MarkoH17** [comment]
Are there any updates on how SSH can be bound to another port, or disabled altogether, after Fedora CoreOS boots?

**Comment by vic1707** [comment]
based on https://www.flatcar.org/docs/latest/setup/security/customizing-sshd/

I included this in my config (the SSH_PORT variable is expanded using envsubst)
```yaml

storage:
  files:
    - path: /etc/ssh/sshd_config.d/00-hardening.conf ## is 00 right?
      mode: 0440
      contents:
        inline: |
          Port ${SSH_PORT}

systemd:
  units:
    - name: sshd.socket
      dropins:
        - name: 10-sshd-port.conf ## is 10 right?
          contents: |
            [Socket]
            ListenStream=${SSH_PORT}
```
And from my *local* testing with qemu, I was able to successfully ssh into fcos on the custom port (and port 22 didn't respond)

Some verification
```
vic1707@telstar:~$ journalctl | grep -i ssh | grep -i listening
Apr 29 23:52:25 telstar systemd[1]: Listening on sshd-unix-local.socket - OpenSSH Server Socket (systemd-ssh-generator, AF_UNIX Local).
Apr 29 23:52:25 telstar systemd[1]: Listening on sshd-vsock.socket - OpenSSH Server Socket (systemd-ssh-generator, AF_VSOCK).
Apr 29 23:52:37 telstar sshd[1257]: Server listening on 0.0.0.0 port XXXXX.
vic1707@telstar:~$ sudo sshd -T | grep port
[sudo] password for vic1707: 
port XXXXX
gatewayports no
```

Is it the way to do it? Is this the best way? Is it perfect?
 - no clue, I'm still learning a lot of things about fcos 😅
